src/rpc/blockchain.cpp

```diff
@@ -198,7 +199,7 @@ static RPCHelpMan getblockcount()
         [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
 {
     LOCK(cs_main);
-    return ::ChainActive().Height();
+    return EnsureChainman(request.context).ActiveChain().Height();
```

Comment:
Agree about `ActiveHeight()` and similar comments below, but may be fine to leave the `LOCK(cs_main)` line as-is since I suspect we'll move to lock annotations for `ActiveChainstate()` in the near future (vs. the nested acquisition in the function definition).

Reply:
I think that'd be a move in the wrong direction. Ideally we'd remove the need to lock `cs_main` completely from the `Active*()` methods. I'm concerned that adding so many additional cs_main locks may cause performance regressions - for example, the optimization in #8007 is no longer possible since cs_main will almost always be locked before calling `IsInitialBlockDownload()`

@jamesob: Did you consider changing the active pointer to being an atomic to remove this requirement as suggested here: https://github.com/bitcoin/bitcoin/pull/19806#issuecomment-768946522?

Reply:
I think you may be forgetting that long before `ChainstateManager` was introduced, [`chainActive` was protected by `cs_main`](https://github.com/bitcoin/bitcoin/blob/0.16/src/validation.h#L445). So unless we dramatically rework the chainstate locking scheme, the `Active{Height,Tip,..}()` methods will always require holding cs_main one way or another. I think ultimately annotations are a cleaner way of doing this because they give callers the flexibility to reuse existing cs_main acquisitions, and avoid recursive locking.

Your suggestion to move chainman's active pointer to an atomic may be a good one, but I think it's largely incidental in terms of performance since we usually need to acquire cs_main anyway when referencing chainstate data. Though my comment above is unrelated to that, since I'm making reference to the chain's protection by cs_main and not ancillary ChainstateManager layout.

[Recent bitcoinperf benchmarks](https://bitcoinperf.com/d/YiV16Vsik/overview?orgId=1&from=now-1y&to=now) don't indicate any regressions in recent revisions, but I will be doing more thorough benchmarking in a bit.